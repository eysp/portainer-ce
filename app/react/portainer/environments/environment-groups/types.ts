import { TagId } from '@/portainer/tags/types';

import {
  TeamAccessPolicies,
  UserAccessPolicies,
  EnvironmentGroupId,
} from '../types';

export interface EnvironmentGroup {
  // Environment(Endpoint) group Identifier
  Id: EnvironmentGroupId;
  // Environment(Endpoint) group name
  Name: string;
  // Description associated to the environment(endpoint) group
  Description: string;
  // List of tags associated to this environment(endpoint) group
  TagIds: TagId[];
  UserAccessPolicies?: UserAccessPolicies;
  TeamAccessPolicies?: TeamAccessPolicies;
}
