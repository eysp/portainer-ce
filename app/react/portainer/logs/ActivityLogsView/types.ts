interface BaseActivityLog {
  timestamp: number;
  action: string;
  context: string;
  id: number;
  username: string;
}
export interface ActivityLogResponse extends BaseActivityLog {
  payload: string;
}

export interface ActivityLog extends BaseActivityLog {
  payload: string | object;
}

export interface ActivityLogsResponse {
  logs: Array<ActivityLogResponse>;
  totalCount: number;
}
