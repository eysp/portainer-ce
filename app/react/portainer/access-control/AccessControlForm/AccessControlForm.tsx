import { FormikErrors } from 'formik';

import { useUser } from '@/portainer/hooks/useUser';

import { FormSectionTitle } from '@@/form-components/FormSectionTitle';
import { SwitchField } from '@@/form-components/SwitchField';

import { EditDetails } from '../EditDetails';
import { ResourceControlOwnership, AccessControlFormData } from '../types';

export interface Props {
  values: AccessControlFormData;
  onChange(values: AccessControlFormData): void;
  hideTitle?: boolean;
  formNamespace?: string;
  errors?: FormikErrors<AccessControlFormData>;
}

export function AccessControlForm({
  values,
  onChange,
  hideTitle,
  formNamespace,
  errors,
}: Props) {
  const { isAdmin } = useUser();

  const accessControlEnabled =
    values.ownership !== ResourceControlOwnership.PUBLIC;
  return (
    <>
      {!hideTitle && <FormSectionTitle>访问控制</FormSectionTitle>}

      <div className="form-group">
        <div className="col-sm-12">
          <SwitchField
            checked={accessControlEnabled}
            name={withNamespace('accessControlEnabled')}
            label="启用访问控制"
            tooltip="启用后，你可以限制对该资源的访问和管理。"
            onChange={handleToggleEnable}
            dataCy="portainer-accessMgmtToggle"
          />
        </div>
      </div>

      {accessControlEnabled && (
        <EditDetails
          onChange={onChange}
          values={values}
          errors={errors}
          formNamespace={formNamespace}
        />
      )}
    </>
  );

  function withNamespace(name: string) {
    return formNamespace ? `${formNamespace}.${name}` : name;
  }

  function handleToggleEnable(accessControlEnabled: boolean) {
    let ownership = ResourceControlOwnership.PUBLIC;
    if (accessControlEnabled) {
      ownership = isAdmin
        ? ResourceControlOwnership.ADMINISTRATORS
        : ResourceControlOwnership.PRIVATE;
    }
    onChange({ ...values, ownership });
  }
}
