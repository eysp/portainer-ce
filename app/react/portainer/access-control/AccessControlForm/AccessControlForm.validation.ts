import { object, string, array, number } from 'yup';

import { ResourceControlOwnership } from '../types';

export function validationSchema(isAdmin: boolean) {
  return object()
    .shape({
      ownership: string()
        .oneOf(Object.values(ResourceControlOwnership))
        .required(),
      authorizedUsers: array(number()),
      authorizedTeams: array(number()),
    })
    .test(
      'user-and-team',
      isAdmin
        ? '你必须至少指定一个团队或用户。'
        : '你必须至少指定一个团队。',
      ({ ownership, authorizedTeams, authorizedUsers }) => {
        if (ownership !== ResourceControlOwnership.RESTRICTED) {
          return true;
        }

        if (!isAdmin) {
          return !!authorizedTeams && authorizedTeams.length > 0;
        }

        return (
          !!authorizedTeams &&
          !!authorizedUsers &&
          (authorizedTeams.length > 0 || authorizedUsers.length > 0)
        );
      }
    );
}
