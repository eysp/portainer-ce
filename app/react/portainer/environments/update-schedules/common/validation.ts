import { array, object, SchemaOf, string, number } from 'yup';

import { parseIsoDate } from '@/portainer/filters/filters';
import { EdgeGroup } from '@/react/edge/edge-groups/types';

import { EdgeUpdateSchedule, ScheduleType } from '../types';

import { nameValidation } from './NameField';
import { typeValidation } from './ScheduleTypeSelector';
import { FormValues } from './types';

export function validation(
  schedules: EdgeUpdateSchedule[],
  edgeGroups: Array<EdgeGroup> | undefined,
  currentId?: EdgeUpdateSchedule['id']
): SchemaOf<FormValues> {
  return object({
    groupIds: array()
      .of(number().default(0))
      .min(1, '至少需要一个组')
      .test(
        '至少一个组必须包含终端',
        (groupIds) =>
          !!(
            groupIds &&
            edgeGroups &&
            groupIds?.flatMap(
              (id) => edgeGroups?.find((group) => group.Id === id)?.Endpoints
            ).length > 0
          )
      ),
    name: nameValidation(schedules, currentId),
    type: typeValidation(),
    scheduledTime: string()
      .default('')
      .test('valid', (value) => !value || parseIsoDate(value) !== null),
    version: string()
      .default('')
      .when('type', {
        is: ScheduleType.Update,
        // update type
        then: (schema) => schema.required('版本为必填项'),
        // rollback
        otherwise: (schema) => schema.required('没有可用的回滚选项'),
      }),
    registryId: number().default(0),
  });
}
