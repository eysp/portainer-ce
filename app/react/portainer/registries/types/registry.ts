import { TLSConfiguration } from '../../settings/types';
import {
  TeamAccessPolicies,
  UserAccessPolicies,
} from '../../environments/types';

export type Catalog = {
  repositories: string[];
};

export enum RegistryTypes {
  ANONYMOUS,
  QUAY,
  AZURE,
  CUSTOM,
  GITLAB,
  PROGET,
  DOCKERHUB,
  ECR,
  GITHUB,
}

export interface RegistryAccess {
  UserAccessPolicies: UserAccessPolicies | null;
  TeamAccessPolicies: TeamAccessPolicies | null;
  Namespaces: string[];
}

export interface RegistryAccesses {
  [key: string]: RegistryAccess;
}

export interface Gitlab {
  ProjectId: number;
  InstanceURL: string;
  ProjectPath: string;
}

export interface Quay {
  UseOrganisation?: boolean;
  OrganisationName: string;
}

export interface Github {
  UseOrganisation: boolean;
  OrganisationName: string;
}

export interface Ecr {
  Region: string;
}

interface RegistryManagementConfiguration {
  Type: RegistryTypes;
  Authentication: boolean;
  Username: string;
  Password: string;
  TLSConfig: TLSConfiguration;
  Ecr: Ecr;
  AccessToken?: string;
  AccessTokenExpiry?: number;
}

export type RegistryId = number;
export interface Registry {
  Id: RegistryId;
  Type: RegistryTypes;
  Name: string;
  URL: string;
  BaseURL: string;
  Authentication: boolean;
  Username: string;
  RegistryAccesses: RegistryAccesses | null;
  Gitlab: Gitlab;
  Quay: Quay;
  Github: Github;
  Ecr: Ecr;
  ManagementConfiguration?: RegistryManagementConfiguration;
}
