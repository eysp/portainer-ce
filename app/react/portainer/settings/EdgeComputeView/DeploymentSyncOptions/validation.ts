import { boolean, number, object, SchemaOf } from 'yup';

import { options as asyncIntervalOptions } from '@/edge/components/EdgeAsyncIntervalsForm';

import { FormValues } from './types';

const intervals = asyncIntervalOptions.map((option) => option.value);

export function validationSchema(): SchemaOf<FormValues> {
  return object({
    EdgeAgentCheckinInterval: number().required('此字段必填。'),
    Edge: object({
      PingInterval: number()
        .required('此字段必填。')
        .oneOf(intervals),
      SnapshotInterval: number()
        .required('此字段必填。')
        .oneOf(intervals),
      CommandInterval: number()
        .required('此字段必填。')
        .oneOf(intervals),
      AsyncMode: boolean().default(false),
    }),
  });
}
