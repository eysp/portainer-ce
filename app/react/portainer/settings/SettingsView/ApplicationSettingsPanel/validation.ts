import { SchemaOf, bool, boolean, number, object, string } from 'yup';

import { isValidUrl } from '@@/form-components/validate-url';

import { Values } from './types';

export function validation(): SchemaOf<Values> {
  return object({
    edgeAgentCheckinInterval: number().required(),
    enableTelemetry: bool().default(false),
    loginBannerEnabled: boolean().default(false),
    loginBanner: string()
      .default('')
      .when('loginBannerEnabled', {
        is: true,
        then: (schema) =>
          schema.required('启用时必须填写登录横幅'),
      }),
    logoEnabled: boolean().default(false),
    logo: string()
      .default('')
      .when('logoEnabled', {
        is: true,
        then: (schema) =>
          schema
            .required('启用时必须提供 Logo URL')
            .test('valid-url', '必须是有效的 URL', (value) =>
              isValidUrl(value)
            ),
      }),
    snapshotInterval: string().required('快照间隔是必填项'),
    templatesUrl: string()
      .default('')
      .test(
        'valid-url',
        '必须是有效的 URL',
        (value) => !value || isValidUrl(value)
      ),
  });
}
