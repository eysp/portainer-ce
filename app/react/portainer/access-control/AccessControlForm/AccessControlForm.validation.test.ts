import { ResourceControlOwnership } from '../types';

import { validationSchema } from './AccessControlForm.validation';

test('when ownership not restricted, should be valid', async () => {
  const schema = validationSchema(true);
  [
    ResourceControlOwnership.ADMINISTRATORS,
    ResourceControlOwnership.PRIVATE,
    ResourceControlOwnership.PUBLIC,
  ].forEach(async (ownership) => {
    const object = { ownership };

    await expect(
      schema.validate(object, { strict: true })
    ).resolves.toStrictEqual(object);
  });
});

test('当所有权受到限制并且没有团队或用户时，应该无效', async () => {
  [true, false].forEach(async (isAdmin) => {
    const schema = validationSchema(isAdmin);

    await expect(
      schema.validate(
        {
          ownership: ResourceControlOwnership.RESTRICTED,
          authorizedTeams: [],
          authorizedUsers: [],
        },
        { strict: true }
      )
    ).rejects.toThrowErrorMatchingSnapshot();
  });
});

test('当所有权受到限制并且用户是管理员时，应该拥有团队或用户', async () => {
  const schema = validationSchema(true);
  const teams = {
    ownership: ResourceControlOwnership.RESTRICTED,
    authorizedTeams: [1],
    authorizedUsers: [],
  };

  await expect(schema.validate(teams, { strict: true })).resolves.toStrictEqual(
    teams
  );

  const users = {
    ownership: ResourceControlOwnership.RESTRICTED,
    authorizedTeams: [],
    authorizedUsers: [1],
  };

  await expect(schema.validate(users, { strict: true })).resolves.toStrictEqual(
    users
  );

  const both = {
    ownership: ResourceControlOwnership.RESTRICTED,
    authorizedTeams: [1],
    authorizedUsers: [2],
  };

  await expect(schema.validate(both, { strict: true })).resolves.toStrictEqual(
    both
  );
});

test('当所有权受到限制时，用户不是团队的管理员，应该有效', async () => {
  const schema = validationSchema(false);

  const object = {
    ownership: ResourceControlOwnership.RESTRICTED,
    authorizedTeams: [1],
    authorizedUsers: [],
  };

  await expect(
    schema.validate(object, { strict: true })
  ).resolves.toStrictEqual(object);
});
