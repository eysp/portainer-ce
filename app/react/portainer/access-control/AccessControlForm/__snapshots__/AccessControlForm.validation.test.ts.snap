// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when ownership is restricted and no teams or users, should be invalid 1`] = `"你必须至少指定一个团队或用户。"`;

exports[`when ownership is restricted and no teams or users, should be invalid 2`] = `"你必须至少指定一个团队。"`;
